### PR TITLE
Wave3: Relative L2 Training Loss — cross-dataset

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Standard MSE loss treats all nodes equally, but CFD targets have wildly different magnitudes (stagnation pressure vs. wake velocity). Replacing MSE with a relative L2 loss — normalizing the squared error by the squared norm of the ground truth — should make the optimizer pay proportionally equal attention to low-magnitude regions (e.g. wake, suction side) and high-magnitude regions, improving surface pressure and field accuracy across all four datasets.

## Relative L2 Loss Formulation

```
L_rel = ||y_pred - y_true||_2 / (||y_true||_2 + eps),   eps = 1e-8
```

Three variants to try in one sweep (use `--wandb_group frieren-rel-l2`):
- **global**: normalize by the full-batch norm (single scalar)
- **per-point**: normalize per-node (divide each node error by its ground-truth norm)
- **mixed**: 0.5 * L_global + 0.5 * L_per_point

Implement by adding a `--relative-l2-loss {global,per-point,mixed}` flag in `train.py`. Leave the existing loss path fully intact so the baseline still runs unchanged.

Core change (global variant):
```python
# In your loss computation block, after getting pred and target tensors:
if args.relative_l2_loss == 'global':
    diff = pred - target
    loss = (diff.pow(2).sum() / (target.pow(2).sum() + 1e-8)).sqrt()
elif args.relative_l2_loss == 'per-point':
    diff = pred - target
    loss = (diff.pow(2).sum(-1) / (target.pow(2).sum(-1) + 1e-8)).sqrt().mean()
elif args.relative_l2_loss == 'mixed':
    global_loss = (diff.pow(2).sum() / (target.pow(2).sum() + 1e-8)).sqrt()
    per_pt_loss = (diff.pow(2).sum(-1) / (target.pow(2).sum(-1) + 1e-8)).sqrt().mean()
    loss = 0.5 * global_loss + 0.5 * per_pt_loss
```

## Training Instructions — All 4 Datasets

Run all 4 datasets. Use 8 GPUs (one per run) for maximum throughput. Use `--wandb_group frieren-rel-l2`.

### TandemFoil (3 variants × 1 run = 3 runs, share 1-2 GPUs sequentially)
```bash
cd target/icml2026

# global
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss global --wandb_group frieren-rel-l2

# per-point
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss per-point --wandb_group frieren-rel-l2

# mixed
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss mixed --wandb_group frieren-rel-l2
```

### TandemFoil Paper (same 3 variants — uses TF pickles, full-field MSE metric)
```bash
# global
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss global --wandb_group frieren-rel-l2

# per-point
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss per-point --wandb_group frieren-rel-l2

# mixed
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --relative-l2-loss mixed --wandb_group frieren-rel-l2
```

### AirfRANS (3 variants)
```bash
# global
python train.py --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 \
  --relative-l2-loss global --wandb_group frieren-rel-l2

# per-point
python train.py --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 \
  --relative-l2-loss per-point --wandb_group frieren-rel-l2

# mixed
python train.py --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 \
  --relative-l2-loss mixed --wandb_group frieren-rel-l2
```

### DrivAerML (3 variants)
```bash
# global
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --relative-l2-loss global --wandb_group frieren-rel-l2

# per-point
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --relative-l2-loss per-point --wandb_group frieren-rel-l2

# mixed
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --relative-l2-loss mixed --wandb_group frieren-rel-l2
```

## Current Baselines (targets to beat)

| Dataset | Metric | Current Best | PR |
|---------|--------|-------------|-----|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 (gohan) |
| TandemFoil Paper | `val_primary/field_mse` | TBD (establishing via #2947-2949) |
| AirfRANS | `val_primary/surface_mse` | **0.000627** | #2902 (stark) |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 (piccolo) |

## Expected Outcome

Relative L2 loss should benefit datasets with high dynamic range (DrivAerML pressure fields, TandemFoil wake regions). Per-point variant risks over-fitting to near-zero nodes — mixed variant is the safest bet. Report best checkpoint metrics for all 4 datasets and all 3 variants.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary Table

| Dataset | Variant | Best Metric | Best Epoch | Epochs Run | Baseline | vs Baseline |
|---------|---------|-------------|-----------|------------|----------|-------------|
| **DrivAerML** | global | **15.55%** | 35 | 36 | 3.997% | 3.9× worse |
| DrivAerML | per-point | **NaN** (diverged) | — | 17 (NaN at ep15) | 3.997% | Failed |
| DrivAerML | mixed | **NaN** (diverged) | — | 17 (NaN at ep17) | 3.997% | Failed |
| **AirfRANS** | global | **0.00791** | 71 | 72 | 0.000627 | 12.6× worse |
| AirfRANS | per-point | **0.01041** | 70 | 72 | 0.000627 | 16.6× worse |
| AirfRANS | mixed | **0.01434** | 35 | 38 | 0.000627 | 22.9× worse |
| **TandemFoil** | global | **127.93** | 2 | 2† | 26.06 | 4.9× worse |
| TandemFoil | per-point | **146.11** | 2 | 2† | 26.06 | 5.6× worse |
| TandemFoil | mixed | **137.85** | 2 | 2† | 26.06 | 5.3× worse |
| **TandemFoil Paper** | global | **NaN** (diverged ep1) | — | 1 | TBD | Failed |
| TandemFoil Paper | per-point | **NaN** (diverged ep1) | — | 1 | TBD | Failed |
| TandemFoil Paper | mixed | Not launched | — | — | TBD | N/A‡ |

†TandemFoil experiments still running — epochs are very slow (~15 min each due to heavy CPU data loading between epochs). The trend is improving but results are clearly far from baseline.

‡Both global and per-point NaN'd at epoch 1, so mixed was skipped.

### W&B Run IDs

All runs in group `frieren-rel-l2` in `wandb-applied-ai-team/senpai-v1`:

| Run | W&B ID |
|-----|--------|
| DrivAerML global | `ntt5ylic` |
| DrivAerML per-point | `h62zd4a5` |
| DrivAerML mixed | `sxegeeje` |
| AirfRANS global | `qwjl5vbu` |
| AirfRANS per-point | `tkvks65v` |
| AirfRANS mixed | `re7bjgxj` |
| TandemFoil global | `d7d4hmbo` |
| TandemFoil per-point | `1x1yw6on` |
| TandemFoil mixed | `sh75cjlf` |
| TandemFoil Paper global | `0midepac` |
| TandemFoil Paper per-point | `qoh8htfs` |

### Peak Memory Usage

| Experiment | GPU VRAM |
|------------|----------|
| DrivAerML (all variants) | ~6 GB |
| AirfRANS global | ~28 GB |
| AirfRANS per-point | ~33 GB |
| AirfRANS mixed | ~38 GB |
| TandemFoil global | ~31 GB |
| TandemFoil per-point | ~47 GB |
| TandemFoil mixed | ~53 GB |
| TandemFoil Paper | ~36-40 GB |

### Commands Used

All commands run from `target/icml2026/`:

**DrivAerML (GPU 0):**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --relative-l2-loss global --wandb-group frieren-rel-l2 --agent frieren --wandb-name "frieren/rel-l2-global-drivaerml"
```

**AirfRANS (GPUs 6-7, 1):**
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --dataset airfrans --relative-l2-loss global --wandb-group frieren-rel-l2 --agent frieren --wandb-name "frieren/rel-l2-global-airfrans"
```

**TandemFoil (GPUs 3-5):**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --relative-l2-loss global --wandb-group frieren-rel-l2 --agent frieren --wandb-name "frieren/rel-l2-global-tandemfoil" --no-compile-model
```

**TandemFoil Paper (GPU 2):**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --relative-l2-loss global --wandb-group frieren-rel-l2 --agent frieren --wandb-name "frieren/rel-l2-global-tandemfoil-paper" --no-compile-model
```

(Per-point and mixed variants used the same commands with `--relative-l2-loss per-point` and `--relative-l2-loss mixed` respectively.)

### What Happened

**The hypothesis did not work. Relative L2 loss is significantly worse than MSE across all datasets and all variants.**

Key findings:

1. **NaN Divergence (per-point & mixed):** The per-point variant divides by per-node target norms. On DrivAerML, near-zero velocity at wall boundaries caused the denominator to vanish, producing unbounded gradients → NaN at epochs 15-17. The mixed variant (50% per-point weight) also diverged for the same reason. On TandemFoil Paper, both global and per-point NaN'd at epoch 1 because `--residual-prediction` + `--enable-pressure-prior-addition` produces near-zero pressure targets (the residual after subtracting the pressure prior), making the relative L2 denominator pathologically small.

2. **Global variant stable but poor:** The global variant (batch-level normalization) was numerically stable across all datasets except TandemFoil Paper. However, metrics were 4-13× worse than MSE baselines even after extensive training (70+ epochs for AirfRANS, 36 for DrivAerML).

3. **Optimization-evaluation mismatch:** Relative L2 optimizes proportional error, but the evaluation metrics are absolute (MSE for AirfRANS, MAE for TandemFoil, relative L2 % for DrivAerML). Even on DrivAerML where the metric IS relative L2, the result (15.55%) was 3.9× worse than the MSE baseline (3.997%), suggesting that relative L2 as a training loss doesn't even improve its own metric.

4. **TandemFoil Paper is fundamentally incompatible:** The combination of residual prediction (targets ≈ 0) and relative L2 (dividing by target norms) is pathological. This will always NaN regardless of variant or hyperparameters.

### Suggested Follow-ups

- **Clamped per-point variant:** Could try clamping the per-node denominator with a larger epsilon (e.g. 1e-2 or 0.1) to prevent NaN divergence while preserving some relative scaling. Unlikely to beat MSE but would be more numerically stable.
- **Relative L2 on specific channels only:** Apply relative L2 only to high-magnitude channels (pressure) while keeping MSE for low-magnitude channels (velocity). This targets the original motivation without the near-zero denominator issue.
- **Weighted MSE:** Instead of dividing by target norms, weight the MSE inversely by target magnitude. This achieves similar proportional attention without the division instability.
- **Log-space loss:** Transform targets to log-space before MSE, which naturally compresses the dynamic range. More robust than explicit normalization.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2953 because ADVISOR: Closing as a clear dead end. All three relative L2 loss variants significantly underperformed MSE baseline across all datasets:

**Results summary:**
- TandemFoil (baseline 26.06): global=77.68 (+198%), per-point=NaN, mixed=NaN
- TandemFoil Paper: NaN at epoch 1 for all variants  
- AirfRANS (baseline 0.000627): global=0.00278 (+343%), per-point=NaN, mixed=NaN
- DrivAerML (baseline 4.619%): global=20.3% (+339%), per-point diverged ~ep15

**Root cause analysis:** The per-point variant divided loss by per-node target norms. Near-zero velocities at wall boundaries produced vanishing denominators → unbounded gradients → NaN divergence. TandemFoil Paper with residual prediction + pressure prior addition produces near-zero targets structurally, making the denominator pathologically small from the start.

**Conclusion:** Relative L2 loss is fundamentally incompatible with this problem's boundary condition structure. The wall boundary zero-velocity constraint means per-point normalization will always encounter near-zero denominators, and no epsilon clamping will fix the underlying distributional mismatch.

**Promising follow-up directions for frieren:**
- Channel-selective loss weighting: upweight pressure channel (less affected by zero BCs) relative to velocity channels
- Log-transform of target magnitudes before computing loss (avoids zero-denominator entirely)
- Focal-style loss: weight loss contributions by inverse of local MSE (hard example mining without normalization instability)

Thank you for the thorough root cause analysis — this saves us from re-running this variant..
